src/Features/LanguageServer/Protocol/Workspaces/LspWorkspaceManager.cs

```diff
@@ -246,35 +270,43 @@ async Task<(Solution Solution, bool IsForked)> GetLspSolutionForWorkspaceAsync(W
+            //
+            //  3. If the cached solution isn't a match, we compare the LSP text to the workspace's text and return the
+            //     workspace text if all LSP text matches. While this does compute checksums, generally speaking that's
+            //     a reasonable price to pay.  For example, we always do this in VS anyways to make OOP calls, and it is
+            //     not a burden there.
             //
-            //   2.  If the cached solution isn't a match, we compare the LSP text to the workspace's text and return
-            //       the workspace text if all LSP text matches. This check is performant as checksums will be computed
-            //       for these documents in order to make requests OOP.  So these are already computed or will be later
-            //       in this request.
+            //  4. Third, we check to see if we have cached a forked LSP solution for the current set of LSP texts
+            //     against the current workspace version. If so, we can just reuse that instead of re-forking and
+            //     blowing away the trees / source generated docs / etc. that we created for the fork.
             //
-            //   3.  Third, we check to see if we have cached a forked LSP solution for the current set of LSP texts
-            //       against the current workspace version. If so, we can just reuse that instead of re-forking and
-            //       blowing away the trees / source generated docs / etc. that we created for the fork.
-            //       
-            //   4.  We have nothing cached for this combination of LSP texts and workspace version.  We have exhausted
-            //       our options and must create an LSP fork from the current workspace solution with the current LSP
-            //       text.
+            //  5. We have nothing cached for this combination of LSP texts and workspace version.  We have exhausted
+            //     our options and must create an LSP fork from the current workspace solution with the current LSP
+            //     text.
             //
             // We propagate the IsForked value back up so that we only report telemetry on forking if the forked
             // solution is actually requested.
 
             // Step 1: Check if nothing has changed and we already verified that the workspace text matches our LSP text.
             if (_cachedLspSolutions.TryGetValue(workspace, out var cachedSolution) && cachedSolution.solution == workspaceCurrentSolution)
-            {
                 return (workspaceCurrentSolution, IsForked: false);
-            }
 
-            // Step 2: Check to see if the LSP text matches the workspace text.
+            // Step 2: Push through any changes to the underlying workspace if it's a mutating workspace.
+            OpenOrEditDocumentsInMutatingWorkspace(workspace);
```

Comment:
This might be something we have to discuss with @tmat and maybe others. This means that a change to a text document isn't applied to the workspace until some _other_ LSP request happens. This is fine if the only way the workspace is ever observed is through LSP; there were still open questions though about what happens if other features (like a debugger) need to request things, and how those might (or might not) synchronize with LSP if those aren't LSP requests.

Reply:
Those features need to go through LSP IMHO.  LSP is really the only way to get the correct view of anything that is opened because LSP owns the opened document text.

Reply:
i agree with David.  Effectively, and *intentionally* we have a system that exposes a 'duality' view of the world.  One view is the one from the PS (which is something that can run entirely outside and independently of a host in the modern world).

The other view is the one from the client, which may only care/consider the open files and their contents.

I would posit that nearly all features (since they are exposed to the user through the client) need to take the LSP view of the world.  So they shoudl come through LSP, go through the LSP server, and then see the LSP solution-snapshot that is totally in sync with the lsp-client files.